### PR TITLE
HPCC-28462 Support resubmit of publisher DFU workunits

### PR DESCRIPTION
**Pull request by afishbeck**
Publisher DFU workunits are parent workunits for all the copy workunits created while publishing a query or packagemap. When some of the child copies fail allow them to be resubmitted by resubmitting the parent publisher DFU workunit.

```
dfuplus action=status server=http://eclwatch:8010 wuid=P20221101-214617
P20221101-214617T2 status: failed - (null) [certification::full_test_distributed]
P20221101-214617T1 status: failed - (null) [certification::full_test_distributed_index]

dfuplus action=resubmit server=http://eclwatch:8010 wuid=P20221101-214617
Resubmitted P20221101-214617

dfuplus action=status server=http://eclwatch:8010 wuid=P20221101-214617
P20221101-214617T2 Finished [certification::full_test_distributed]
P20221101-214617T1 Finished [certification::full_test_distributed_index]

```
Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dali/dfu/dfuwu.cpp [hunk 1]
```diff
@@ -416,6 +416,87 @@ static bool notePublisherSubTaskState(const char *subtaskWuid, DFUstate state)
     return true;
 }
 
+static bool prepareToResubmitFailedPublisherTasks(const char *publisherWuid, StringArray &taskWuidsToResubmit)
+{
+    //make sure wuid is a publisher, but not a child task of a publisher
+    if (!publisherWuid || 'P'!=*publisherWuid || strchr(publisherWuid, 'T')!=nullptr)
+       throw MakeStringException(-1,"DFUWU: Attempt to resubmit publisher workunit with non publisher wuid %s", publisherWuid);
```
Comment (ghalliday) [inline dali/dfu/dfuwu.cpp:423]: trivial: indented by 3 rather than 4.  Same for lines below.

File: dali/dfu/dfuwu.cpp [hunk 2]
```diff
@@ -416,6 +416,87 @@ static bool notePublisherSubTaskState(const char *subtaskWuid, DFUstate state)
...
+    progress->setPropInt("Edition", ++edition);
+    conn->commit();
+    return true;
+}
+
+static bool resubmitFailedPublisherTasks(const char *publisherWuid)
+{
+    StringArray taskWuidsToResubmit;
+    if (!prepareToResubmitFailedPublisherTasks(publisherWuid, taskWuidsToResubmit))
+        return false;
```
Comment (ghalliday) [inline dali/dfu/dfuwu.cpp:494]: trivial: never returns false, but fine as it stands.

**Review by ghalliday — APPROVED** [review]
@afishbeck Will merge as it stands.

File: dali/dfu/dfuwu.cpp (hunk 1 above)
Comment (jakesmith) [reply dali/dfu/dfuwu.cpp:423]: trivial: makeStringException/makeStringExceptionV preferred over MakeStringException

(and others below)

File: dali/dfu/dfuwu.cpp [hunk 3]
```diff
@@ -416,6 +416,87 @@ static bool notePublisherSubTaskState(const char *subtaskWuid, DFUstate state)
     return true;
 }
 
+static bool prepareToResubmitFailedPublisherTasks(const char *publisherWuid, StringArray &taskWuidsToResubmit)
+{
+    //make sure wuid is a publisher, but not a child task of a publisher
+    if (!publisherWuid || 'P'!=*publisherWuid || strchr(publisherWuid, 'T')!=nullptr)
```
Comment (jakesmith) [inline dali/dfu/dfuwu.cpp:422]: trivial: use isDfuPublisherWuid() for clarity?

**Review by jakesmith — APPROVED** [review]
@afishbeck - a couple of trivials, but looks good/approved

**Pull request merged**